### PR TITLE
Added version spec ability for dub commands

### LINKED ISSUE #1979: `dub describe` can't describe specific package version
I think it must be like `dub describe mypackage@1.2.3`
related to https://github.com/mesonbuild/meson/issues/7496

### PR DESCRIPTION
**Pull request by deviator**
fix #1979
ping @andre2007

### CODE REVIEW

**Comment by andre2007** [comment]
Thanks for the additional tests. The changes looks good for me.

**Comment by andre2007** [comment]
This pr might also solve the main issue of this 
https://github.com/dlang/dub/issues/1751
(Other issue is, wheter `dub run mainPackage:subPackage@version` is working)

**Comment by deviator** [comment]
ping @Geod24

File: source/dub/commandline.d [hunk 1]
```diff
@@ -898,7 +898,7 @@ class InitCommand : Command {
 			p.description = input("Description", p.description);
 			p.authors = input("Author name", author).split(",").map!(a => a.strip).array;
 			p.license = input("License", p.license);
-			string copyrightString = .format("Copyright © %s, %-(%s, %)", Clock.currTime().year, p.authors);
+			string copyrightString = format("Copyright © %s, %-(%s, %)", Clock.currTime().year, p.authors);
```
Comment (Geod24) [inline source/dub/commandline.d:901]: Unrelated

File: test/.gitignore [hunk 2]
```diff
@@ -12,5 +12,8 @@ custom-unittest/custom-unittest
 path-subpackage-ref/test
 subpackage-ref/test
 subpackage-common-with-sourcefile-globbing/mypackage*
+version-spec/**/CMakeLists.txt
+version-spec/**/foo.cmake
+version-spec/**/foo
 
-/test_registry
+/test_registry
```
Comment (Geod24) [inline test/.gitignore:19]: Looks accidental too

File: test/version-spec.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+[[ $($DUB build -n foo | head -1) == *"/newfoo/" ]] || false
+[[ $($DUB build -n foo@1.0.0 | head -1) == *"/newfoo/" ]] || false
+[[ $($DUB build -n foo@0.1.0 | head -1) == *"/oldfoo/" ]] || false
+
+[[ $($DUB run -n foo | tail -1) == 'new-foo' ]] || false
+[[ $($DUB run -n foo@1.0.0 | tail -1) == 'new-foo' ]] || false
+[[ $($DUB run -n foo@0.1.0 | tail -1) == 'old-foo' ]] || false
+
+$DUB remove-local "$CURR_DIR/version-spec/newfoo"
+$DUB remove-local "$CURR_DIR/version-spec/oldfoo"
```
Comment (Geod24) [inline test/version-spec.sh:33]: Newline

File: test/version-spec/newfoo/dub.sdl [hunk 4]
```diff
@@ -0,0 +1,3 @@
+name "foo"
+version "1.0.0"
+targetType "executable"
```
Comment (Geod24) [inline test/version-spec/newfoo/dub.sdl:3]: Newline

File: test/version-spec/oldfoo/dub.sdl [hunk 5]
```diff
@@ -0,0 +1,3 @@
+name "foo"
+version "0.1.0"
+targetType "executable"
```
Comment (Geod24) [inline test/version-spec/oldfoo/dub.sdl:3]: Newline

File: source/dub/commandline.d (hunk 1 above)
Comment (deviator) [reply source/dub/commandline.d:901]: agree, it's unrelated, but in other place it was a question https://github.com/dlang/dub/pull/1980#issuecomment-665205310, I was think it can be fixed in this place too (I rewert it in next commit)

**Comment by deviator** [comment]
@Geod24 
> ```
> [...]
>   fetch <name>[@<version-spec>]
>                         Manually retrieves and caches a package
>   add <package>[@<version-spec>] [<packages...>]
>                         Adds dependencies to the package file.
> ```

Question 1: `fetch` and `remove` commands has `<name>` parameter, but `add` and many other commands has `<package>` parameter, those have different meaning? May be in all places need use `<package>`?
 
> `dub remove` supports a `--version` argument, which is now redundant. It should eventually be deprecated, however that can go in another PR.

Question 2: deprecation must be in help or when use or both? is needed deprecation at compilation?

**Comment by wilzbach** [comment]
> Question 1: fetch and remove commands has <name> parameter, but add and many other commands has <package> parameter, those have different meaning? May be in all places need use <package>?

Yes, <package> for all sounds good.

> Question 2: deprecation must be in help or when use or both? is needed deprecation at compilation?

It can be removed entirely from the help (that's not even considered as a deprecation). The deprecation, however, is a warning when the argument is supplied to dub by the user.

File: test/.gitignore (hunk 2 above)
Comment (deviator) [reply test/.gitignore:19]: cmake files and `foo` binary generates by `version-spec.sh` test, why it don't be in .gitignore?

**Comment by deviator** [comment]
@wilzbach @Geod24 help updated, `--version` parameter hidden from help, warning added when user use it

File: test/.gitignore (hunk 2 above)
Comment (wilzbach) [reply test/.gitignore]: The comment was about the removed newline.

**Comment by deviator** [comment]
ping @wilzbach @Geod24

**Review by Geod24 — APPROVED** [review]
Amazing, thanks!

**Comment by Geod24** [comment]
Aaaaaand we forgot about the changelog, once again. Also this deprecation triggers everywhere in the test-suite, and I think it sometimes does when it shouldn't (or with a bad error message). Will submit a PR soon.

File: test/version-spec.sh [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+[[ $($DUB generate cmake foo@1.0.0 | head -1) == *"/newfoo/" ]] || false
+[[ $($DUB generate cmake foo@0.1.0 | head -1) == *"/oldfoo/" ]] || false
+
+[[ $($DUB build -n foo | head -1) == *"/newfoo/" ]] || false
+[[ $($DUB build -n foo@1.0.0 | head -1) == *"/newfoo/" ]] || false
+[[ $($DUB build -n foo@0.1.0 | head -1) == *"/oldfoo/" ]] || false
+
+[[ $($DUB run -n foo | tail -1) == 'new-foo' ]] || false
+[[ $($DUB run -n foo@1.0.0 | tail -1) == 'new-foo' ]] || false
+[[ $($DUB run -n foo@0.1.0 | tail -1) == 'old-foo' ]] || false
```
Comment (Geod24) [inline test/version-spec.sh:30]: What was the intent of those `|| false` ? They don't do anything. `X || false` will be `X` no matter what.

Comment (deviator) [reply test/version-spec.sh:30]: I not good in bash, but lines 12, 13, 14, 24, 25, 26 (where used regexp like `*"somestring"`) without `|| false` prints `[ERROR] sh:12 command failed`

full output from my machine:
```
Package is already registered: foo (version: 1.0.0)
Package is already registered: foo (version: 0.1.0)
[ERROR] sh:12 command failed <<<<-- this lines
[ERROR] sh:13 command failed
[ERROR] sh:14 command failed
[ERROR] sh:24 command failed
[ERROR] sh:25 command failed
[ERROR] sh:26 command failed
Deregistered package: foo (version: 1.0.0)
Deregistered package: foo (version: 0.1.0)
Fetching dub 1.9.0...
Fetching dub 1.10.0...
Removing dub in /home/deviator/.dub/packages/dub-1.9.0/dub/
Removed package: 'dub'
Removed dub, version 1.9.0.
Removing dub in /home/deviator/.dub/packages/dub-1.10.0/dub/
Removed package: 'dub'
Removed dub, version 1.10.0.
```
but test passes, in all lines `dub` works correctly... I don't understand why it prints and why it not break test at all.
I think this output wasty and mislead. For other lines `|| false` was added for consistent.
